### ISSUE TITLE
Improve content blocking to match Safari + 3rd party content blockers

### ISSUE THREAD

**Issue by digitalqualms** [mozilla-mobile/firefox-ios#7374]
### Why/User Benefit/User Problem

Hi,

Content blocking in Firefox on iOS is not as good as Safari with 3rd party content blockers (e.g. AdGuard). I would switch my default browser if Firefox were able to perform as well in this regard.

Perhaps do one or more of the following:
1. Inherit Safari content blockers somehow (I'm guessing there are currently API restrictions against this. Maybe an iOS developer can file more specific feedback to Apple on what would be required)
2. Use more comprehensive block lists (e.g. what AdGuard and or other dedicated Safari content blockers use) by default. It looks like Edge for iOS ships with Adblock Plus but it's not as good as Safari + 3rd party either
3. Add more customization options for content blocking (perhaps a slider with how strict it is)

Safari (iOS 14.0) + AdGurard (4.0.3) + Firefox Focus (8.1.5):
<img src="https://user-images.githubusercontent.com/50979043/93970871-96e5f100-fd34-11ea-9d15-7338f5e14a8a.PNG" width="33%">

Firefox (28.2)
<img src="https://user-images.githubusercontent.com/50979043/93971108-0a87fe00-fd35-11ea-8e74-ac38eb97f934.PNG" width="33%">

Edge (45.8.14)
<img src="https://user-images.githubusercontent.com/50979043/93971221-46bb5e80-fd35-11ea-9241-a6763e4cbe84.PNG" width="33%">

### Acceptance Criteria (how do I know when I’m done?)

Ad blocking is nearly as good as Safari + 3rd party blockers

### Not included

Thanks for considering this!

**Comment by athomasmoz** [comment]
Thanks for logging @digitalqualms . We have a Strict Enhanced Tracking Protection (under Settings > Tracking Protection) that blocks tracking content. Have you tried that and does it cover what you're looking for? I'm wondering if the request should be to add more content blocking or to make the strict option more prominent in the UI

**Comment by digitalqualms** [comment]
Good question. Strict Enhanced Tracking Protection seems to be an improvement and closer to what I hoped to see. I think I overlooked it because I didn't realize "Tracking Protection" applied to blocking visible content - I expected that menu to be just cookies, tracking pixels, etc. It's clear once you read the description of Strict Enhanced Tracking Protection that it does block content.

I'll give it a shot but I'd still like to have the powerful customization that Safari is able to provide (again, I realize Apple may have to make concessions/changes to allow for that).

**Comment by mikelambrellis** [comment]
I'm with digitalqualms. Even though I use Firefox everywhere else, I'd prefer the adblocking to be improved before I switch default browser on iOS. With respect to testing and acceptance criteria, I'd suggest canyoublockit.com. At the moment FF on iOS is worse than Safari or Edge.

**Comment by bcutter** [comment]
Copy that. Since Apple allows to change default bewieset with iOS -4 I finally gave Firefox a chance to prove over few weeks. The first thing which is a massive CON compared to Safari is the huge amount of ads I have to look at when surfing on the mobile broadband (I use a network based ad blocker in Home WiFi).

Firefox without improved adblocking abilities is a clear NO for me. Nice browser but not good enough. I guess it's because of some Apple limitations right?

Anyway, strongly vote for this feature request.

**Comment by dmose** [comment]
I have a different use case: I'm using https://freedom.to to help block out distractions, and on iOS it uses a content blocker in Safari to help manage this.  I'm hoping to avoid having to uninstall Firefox on iOS in order to take back my attention from my devices.

**Comment by pshirshov** [comment]
> Anyway, strongly vote for this feature request.

A lot of people shared the same concern and/or requested a good content blocker. Though for some reason FF team never acknowledges the issue and ignores all the arguments. The reason is unclear and this situation creates a breeding ground for wild conspiracy theories.

**Comment by DarrenSchwartz** [comment]
Agree.  Mozilla should recognize the importance of this.  Firefox is a great experience on every other platform, but the ecosystem effect is ruined because of the lack of ad blocking on ios.  I would love to give up safari and use Firefox 100% everywhere, but like the others here, I find the browsing experience in safari to be superior because of good content blocking.  Firefox on ios will not have a chance at being successful until it has ad blocking parity with safari.  Sorry, but it's true.

**Comment by athomasmoz** [comment]
Thank you for the input. We have a Strict Enhanced Tracking Protection (under Settings > Tracking Protection) that blocks tracking content, meaning a lot less ads will show. We used to have this on by default, but it can result in website breakage, so we turned it off by default mid-2019 and left the option to turn it on. I'm curious if your comments about the number of ads are with strict turned on or in our standard mode. This will help me to determine if making strict ETP more prevalent could be a solution

**Comment by DarrenSchwartz** [comment]
Ok, thanks for the clarification.  I had standard tracking protection (not strict).  I have changed it now and will have to evaluate for some time to see if it is comparable to safari.
I think part of this confusion comes from referring to this function as tracking protection instead of ad blocking or content blocking.  People think of tracking differently from the display of annoying ads (words matter).  Tracking can be invisible, while ads are very visual.  If the strict tracking protection really does work as well as safari at blocking visual ads, then you can probably resolve this complaint with clearer language around the feature.  Maybe even a separate switch that clearly says "block ads" would help the user understand what it does.  The picture below is from desktop Firefox, but illustrates the point.  All it talks about is tracking and nowhere does it mention ads.  I know that mobile Firefox does mention ads in the description, but calling it tracking protection is the root of the problem. People have come to know the difference between tracking and ads.  They want to know that their preference to block visual ads is being explicitly understood.  I'll report back my experience with strict tracking protection.  As a side note, it would be fantastic if we could manage custom lists used by the blocker (like uBlock Origin).

![Untitled picture](https://user-images.githubusercontent.com/33355526/103931360-0958c200-50d5-11eb-915f-621892334ca3.png)

**Comment by pshirshov** [comment]
> Thank you for the input. We have a Strict Enhanced Tracking Protection (under Settings > Tracking Protection) that blocks tracking content, meaning a lot less ads will show. 
Not really. Maybe on some sites Strict Mode works well, but I would say that I see about 80% of ads on the sites I usually visit. Which is a lot comparing to 0-10% with uBlock or AdGuard.

**Comment by DarrenSchwartz** [comment]
So I went to msn news and google news with strict tracking protection on.  It did indeed block all of the visual ads I came across, but most of the elements were not collapsed (hidden) like they were in safari.  So it's better than nothing, but safari was a far better experience because it wasn't full of empty white space.  Here are a few examples of issues on firefox that did not exist on safari:

**msn news**
- "add to home screen" popup on initial page load (the one that so helpfully points to the button for you).
- opening an article presents the "view in app" banner that you have to dismiss every time.   I tried reddit too because that site is downright abusive with that "view in app" garbage, but in that case, neither browser blocked the popup.

**google news**
- firefox allowed a sticky banner at the bottom of the page where safari did not (it was empty, but again, not collapsed).
- again, interstitial ads had their content blocked, but the element was not collapsed.
- neither browser blocked all of 8 social buttons, but safari blocked half of them.

For some reason neither one is blocking autoplay video even though it is turned off in ios settings.

For the record, the ios blockers I have installed are: adblock plus, firefox focus, purify, and 1blocker legacy.

**Comment by andreiburov** [comment]
Content blockers like Adguard make it possible to turn on filtering lists for ads that are constantly maintained. I definitely support adding a content blockers support to FF.

Strict mode on FF does not block plenty of ads, while at the same time this mode breaks embeddings, i.e. twitter posts embedded in a blog page.

**Comment by dareducky** [comment]
I would love to see a proper adblock on firefox. 
As it stand, firefox on iOS is a horrible user experience. Here are some quick comparisons between third party browsers on iOS (as to quickly dismiss any argument of third party browsers are too limited on iOS, which is true but not relevant to this issue).

- **Edge with adblock plus:**

![3C1FFB34-3036-40EB-B89E-71571CF277C0](https://user-images.githubusercontent.com/76420385/116411576-39ef7380-a80c-11eb-8628-35f810f4ad34.jpeg)

- **iCab with content filter enabled:**

![C2ABD510-5D6B-4555-B577-35B0C8C544F5](https://user-images.githubusercontent.com/76420385/116411593-3cea6400-a80c-11eb-8dc5-649f1ce0f896.jpeg)

- **Firefox with strict tracking blocking + Nextdns filtering:**

![01BB4226-42E2-4B07-A199-A00907A5D152](https://user-images.githubusercontent.com/76420385/116411612-407deb00-a80c-11eb-90c3-28d16b2e7ca4.jpeg)

This is just one exemple, if I had time I could parse cookies notices and many ads that firefox simply fail to block.

**Comment by mcl21014** [comment]
Edge for iOS, at least in the preview, has Adblock Plus baked in. I do not like ABP as they take money for questionable "acceptable ads", but this is competition to Firefox. Edge also has its own tracking protection and third party cookie blocking as an option (another issue). Ublock origin likely will never come, see the developer's reasoning as Apple has made it hard to block ads via any extension on desktop Safari, and even though this issue is different, if I were him I wouldn't want to work with iOS either, even in Firefox. That doesn't mean Mozilla cannot port uBlock Origin code or create their own adblocker for iOS, although the latter may be more difficult, not too sure.

**Comment by bcutter** [comment]
It's just damn annoying to see all that ads and consent/cookie banners when using Firefox on iOS outside of a network ad-blocked WiFi. Until Firefox significantly improves here, that's a no-go for Firefox and a Go for Safari. Sorry to say but that's my two cents. And there's no progress here, not at all, unfortunately.

**Comment by rogeriochaves** [comment]
I’m with @bcutter, people usually mention ads but what I really can’t stand is the cookie consent banners, I’m in Europe so EVERY website needs a click extra, in a different place, to let me access it, I cannot use the web without getting annoyed

It’s not enough just to block ads, other content blockers goes beyond that also blocking cookie banners, install my app banners, etc, because of that I will have to go with Safari for now

**Comment by c-thun** [comment]
Adding on to this, just blocking ads does not help save the wasted screen real estate always, for example as in . This will need implementation of cosmetic filters as well so as to match the experience provided by Safari. Until that happens, using Firefox feels worse than using Safari + AdGuard

**Comment by rogeriochaves** [comment]
For now I've done my own build of firefox-ios, bundling https://www.i-dont-care-about-cookies.eu/ in it, so I could go back to use it instead of safari. Here is [the repo](https://github.com/rogeriochaves/firefox-ios-i-dont-care-about-cookies), and [the blogpost](https://rchaves.app/firefox-ios-without-cookies/)

**Comment by bcutter** [comment]
Wow. If you can - why can't Mozilla?

**Comment by captn3m0** [comment]
Orion Browser comes with uBlock origin supported, and plans to support every extension in the Chrome and Firefox stores: https://blog.kagi.com/orion-features

How is this labelled P2! ¯\\_(ツ)_/¯

**Comment by pshirshov** [comment]
Apple-only browser? Really? Seriously?

**Comment by DarrenSchwartz** [comment]
> Orion Browser comes with uBlock origin supported, and plans to support every extension in the Chrome and Firefox stores: https://blog.kagi.com/orion-features

Looks cool, but this thread is about Firefox for ios, not Mac. There is still no adequate ad blocking or element hiding on Firefox for ios. I really want to move to Firefox on ios because I use it everywhere else with uBlock Origin, but the ad situation is intolerable. I get that ios has limitations, but surely something can be done to pull in blocklist rules. Is there not a way to take advantage of Safari's underlying use of content blockers?

**Comment by captn3m0** [comment]
Orion is supported on iOS and Mac: 

> Currently Orion supports large number of popular extensions on Desktop with support for more to come. Orion even supports some extensions out of the box on iOS, but we have much more to do here.

Testflight Link: https://testflight.apple.com/join/DeC8ZDnu

**Comment by ahydronous** [comment]
It blows my mind how Mozilla is still fiercely opposing this when you basically cheerily patched in uBlock Origin support in Firefox Preview at the earliest possible moment.

iOS users deserve the same amount of protection as Android users. Not giving them an adequate adblocker is akin to leaving them out in the cold.

Ads are scam / attack vectors and I've personally gotten all my less tech-literate family and friends off of Firefox because of this. That's at least 10 users gone. A few of them even asked if they should tell their friends.

**Comment by mr-manuel** [comment]
Why is Mozilla not able to block ads like Brave do? When I watch YouTube on the iPad I have to use Brave to get rid of all the ads. With Firefox it‘s a mess. Even, if I set the tracking protection to strict.
So there is no „because iOS“ excuse for me.

**Comment by Prurite** [comment]
I heavily use Firefox for desktop, but the experience on ipad is just _frustrating_. I’m now hang in the middle, with both Safari and Firefox unsatisfactory. Not to mention other addons, at least we should get the content blocking feature aligned with other browsers.

**Comment by bcutter** [comment]
I also don't get why the "slide to left/right over the address bar" does not switch tabs like it does for Safari on iOS as well as for Firefox on Android.

Off-topic, but it adds another puzzle to the feeling "Firefox on iOS has ZERO priority and/or attention."

So annoying, especially as nothing happens. Not a single reaction or time schedule, just nothing.

**Comment by maikol-solis** [comment]
Today I discovered this website https://libredns.gr/ which allows adding a system-wide configuration to block ads on iOS. It isn't perfect, but it does the job.

**Comment by maikol-solis** [comment]
@no-milestone Ok I got it. But what is the alternative that Firefox offers us to avoid this bad practice?

**Comment by scineram** [comment]
It's unreal that UBOL works in Safari but not Firefox. What are you doing? What are you paid for?

**Comment by icurl** [comment]
Has there been any updates and/or progression on this? Content blocking (or ad blocking in layman's terms) would truly make Firefox the absolute go-to for iOS web browsing. Firefox Focus (iOS) is very much capable of this, making its absence as an opt-in function for the regular app quite peculiar. 

I really hope to see it implemented, and I'd like to give my thanks to the gracious maintainers and contributors who have worked on the iOS app thus far! 😀

**Comment by Sevilum** [comment]
I want to chime in as a EU-resident. Are there any plans to add support for firefox extensions on iOS (which would also include adblockers) since browsers can now run on their own engines and not just WebKit?

I really love Firefox, and browser extensions on mobile (iOS) would be such a great improvement.

**Comment by captn3m0** [comment]
Existing issue for moving from Webkit to Gecko: https://github.com/mozilla-mobile/firefox-ios/issues/19063

**Comment by iaxat** [comment]
Even Brave has integrated ad-blocking for iOS which is way better than Safari or anything out there.

**Comment by bcutter** [comment]
I left Firefox on iOS years ago mainly because of this (plus a not that intuitive UI/UX compared to Safari).

**Comment by scineram** [comment]
Why didn't they sue Apple in EU for monopoly abuse?

**Comment by maikol-solis** [comment]
We need a proper ad blocker on iOS, please.

**Comment by allagashSage** [comment]
It is super silly that in year of 2026, Firefox on iOS does not have a proper content blocker. Even Brave figured this out by building its own. It prevents folks (at least me) using FF as a primary browser due to this missing feature. So, I have to use Safari with Adguard and then iOS Firefox to keep my bookmarks properly organized. Will Firefox ever heed this call? Only time will tell.

**Comment by iMe** [comment]
Is there any fork of iOS firefox with an ad blocker build in?

**Comment by BasixKOR** [comment]
I believe this can be closed now: https://support.mozilla.org/en-US/kb/block-ads-firefox-ios

**Comment by allagashSage** [comment]
Yeah, and then promptly open a new one to improve on what was implemented - add more filters lists and allow users to add their own.

**Comment by scineram** [comment]
Disgracefully Ublock Origin is only effective in Safari, not other browsers. If only that restriction was illegal.
